### PR TITLE
Migrate the Amazon provider to ASP.NET Core 2.0

### PR DESCRIPTION
**Pull request by martincostello**
Migrate the Amazon provider to ASP.NET Core 2.0.

Relates to #170.

### CODE REVIEW

File: src/AspNet.Security.OAuth.Amazon/AmazonAuthenticationHandler.cs [hunk 1]
```diff
@@ -4,31 +4,37 @@
...
+            [NotNull] ISystemClock clock)
+            : base(options, logger, encoder, clock)
         {
         }
 
         protected override async Task<AuthenticationTicket> CreateTicketAsync([NotNull] ClaimsIdentity identity,
             [NotNull] AuthenticationProperties properties, [NotNull] OAuthTokenResponse tokens)
         {
-            var endpoint = Options.UserInformationEndpoint;
+            string endpoint = QueryHelpers.AddQueryString(Options.UserInformationEndpoint, "access_token", tokens.AccessToken);
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.Amazon/AmazonAuthenticationHandler.cs]: Why this change? We don't want functional changes as part of the 2.0 port.

File: src/AspNet.Security.OAuth.Amazon/AmazonAuthenticationOptions.cs [hunk 2]
```diff
@@ -20,28 +21,30 @@ public class AmazonAuthenticationOptions : OAuthOptions
...
         /// <summary>
         /// Gets the list of fields to retrieve from the user information endpoint.
         /// </summary>
-        public ISet<string> Fields { get; } = new HashSet<string>
-        {
-            "email",
-            "name",
-            "user_id"
-        };
+        public ICollection<string> Fields { get; } = new HashSet<string>();
```
Comment (kevinchalet) [inline src/AspNet.Security.OAuth.Amazon/AmazonAuthenticationOptions.cs]: Revert.

**Review by kevinchalet — CHANGES_REQUESTED** [review]

File: src/AspNet.Security.OAuth.Amazon/AmazonAuthenticationHandler.cs (hunk 1 above)
Comment (martincostello) [reply src/AspNet.Security.OAuth.Amazon/AmazonAuthenticationHandler.cs]: This comes from how I originally implemented Amazon authentication in my own project before I did the initial PR to add it here: [link](https://github.com/martincostello/alexa-london-travel-site/blob/28b5adea435fdaed519add79e4f4dff1fab30889/src/LondonTravel.Site/Identity/Amazon/AmazonHandler.cs#L33)

File: src/AspNet.Security.OAuth.Amazon/AmazonAuthenticationOptions.cs (hunk 2 above)
Comment (martincostello) [reply src/AspNet.Security.OAuth.Amazon/AmazonAuthenticationOptions.cs]: Why? It's functionally equivalent because the fields are populated by the constructor now instead of inline and `ISet<T>` wasn't found when I tried to initially compile it for 2.0.

File: src/AspNet.Security.OAuth.Amazon/AmazonAuthenticationHandler.cs (hunk 1 above)
Comment (kevinchalet) [reply src/AspNet.Security.OAuth.Amazon/AmazonAuthenticationHandler.cs]: It's always preferable to flow the access token via the `Authorization` header, if possible (which is the approach recommended by the OAuth2 bearer spec).

If Amazon deprecated or removed that, please open a new ticket so we can change the 1.x provider too.

File: src/AspNet.Security.OAuth.Amazon/AmazonAuthenticationOptions.cs (hunk 2 above)
Comment (kevinchalet) [reply src/AspNet.Security.OAuth.Amazon/AmazonAuthenticationOptions.cs]: Because it's not a necessary change and it makes the 2.0 signature inconsistent with what we have in the 1.x version. Only changes that are strictly necessary to make the providers work with 2.0 are accepted as part of the "2.0 port" PRs.

**Comment by martincostello** [comment]
@PinpointTownes Commit updated to revert those changes.

**Comment by kevinchalet** [comment]
Thanks. Looks like you forgot to remove the helper class, but I'll go ahead and remove it before merging your PR.

**Review by kevinchalet — APPROVED** [review]

**Comment by kevinchalet** [comment]
Merged, thanks! https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/commit/03cf0f4f69617429bbc0b39b714e235b5fc71bee
